src/pcapi/core/offers/repository.py

```diff
@@ -275,3 +276,34 @@ def get_sold_out_offers_count_for_venue(venue_id) -> int:
     query = Offer.query.filter(Offer.venueId == venue_id)
     query = _filter_by_status(query, datetime_now, SOLD_OUT_STATUS)
     return query.distinct(Offer.id).count()
+
+
+def get_and_lock_stock(stock_id: int) -> Stock:
+    """Returns `stock_id` stock with a FOR UPDATE lock
+    Raises StockDoesNotExist if no stock is found.
+    WARNING: MAKE SURE YOU FREE THE LOCK (with COMMIT or ROLLBACK) and don't hold it longer than
+    strictly necessary.
+    """
+    # Use `with_for_update()` to make sure we lock the stock while perfoming
+    # the booking checks and update the `dnBookedQuantity`
+    # This is required to prevent bugs due to concurent acces
+    # Also call `populate_existing()` to make sure we don't use something
+    # older from the SQLAlchemy's session.
+    stock = Stock.query.filter_by(id=stock_id).populate_existing().with_for_update().one_or_none()
+    if not stock:
+        raise StockDoesNotExist()
+    return stock
+
+
+def check_stock_consistence() -> List[Stock]:
+    return (
+        Stock.query.outerjoin(Stock.bookings)
+        .group_by(Stock.id)
+        .having(
+            Stock.dnBookedQuantity
+            != func.coalesce(
+                func.sum(Booking.quantity).filter(or_(Booking.isCancelled == False, Booking.isCancelled == None)), 0
```

Comment:
`Booking.isCancelled` n'est pas nullable. On devrait donc pouvoir écrire :


```suggestion
                func.sum(Booking.quantity).filter(Booking.isCancelled.is_(False)), 0
```

Reply:
en fait, oui mais non.
`isCancelled` n'est normalement pas nullable sauf dans le cas où l'on n'a pas de booking pour un stock, auquel cas il apparait comme `null` à cause de la jointure externe à droite (je ne fournis pas l'aspirine qui va avec).

Reply:
Je vais quand même vérifier ça, mais dans mes essais, il me semblait que c'était nécessaire.

Reply:
Manifestement, c'était un reste conservé lors de la construction de la requête et qui n'est plus nécessaire -> exit